### DISCUSSION TITLE
accessing area type and region vocabularies

### DISCUSSION THREAD

**Discussion by taylor13** [cf-convention/discuss#397] [Q&A about using CF]
### Question

From a code, say written in python, can one directly read the list of area types (found [here](http://cfconventions.org/Data/area-type-table/current/build/area-type-table.html)) or the list of regions (found [here](http://cfconventions.org/Data/standardized-region-list/standardized-region-list.current.html))?  I see they are displayed as html files, but is that built on a reference source, or is that the reference source.  

We're writing a code that needs to check the input provided by a user against the CF controlled vocabularies. 

I guess I'd be curious also how one would check whether a standard_name provided by a user is in the CF list of standard names.

**Comment by PythonCHB** [comment]
The html is not the source, though I’m not CC sure where the actual source is. 

I’m sure someone else can answer your question directly, but there are at least a couple cf-checkers available:

https://github.com/cf-convention/cf-convention.github.io/blob/main/software.md

I’d take a look at their code to see where they are getting the vocabularies.

**Comment by taylor13** [comment]
Nice lead, @PythonCHB.  In the checker the paths are given as:

STANDARDNAME = 'http://cfconventions.org/Data/cf-standard-names/current/src/cf-standard-name-table.xml'
AREATYPES = 'http://cfconventions.org/Data/area-type-table/current/src/area-type-table.xml'
REGIONNAMES = 'http://cfconventions.org/Data/standardized-region-list/standardized-region-list.xml'

And the xml's are fairly simply structured.  For example, here's the beginning of the area types xml:
```
<area_type_table xsi:noNamespaceSchemaLocation="area-type-table-1.1.xsd">
    <title>Area Type Table</title>
    <version_number>11</version_number>
    <date>06 July 2023</date>
    <institution>Centre for Environmental Data Analysis</institution>
    <contact>support@ceda.ac.uk</contact>
    <entry id="air">
        <description>
            The area type of "air" distinguishes the atmospheric portion of a domain from the land portion (or any
            other non-air portion). It is needed when a variable such as wind speed is reported, for example, on a 
            surface at mean sea level  pressure. At most locations over land, this surface would be underground 
            and would need to be excluded from consideration, which could be indicated by specifying a 
            cell_ methods "where air".
        </description>
    </entry>
    <entry id="all_area_types">
        <description/>
    </entry>
<entry id="bare_ground">
<description/>
```
So I suppose, you just need to ingest the xml.   Have I got that right?  Do we all agree that those xml's are the reference controlled vocabularies?

**Reply by davidhassell** [reply]
> those xml's are the reference controlled vocabularies?

That is indeed my assumption.

**Reply by japamment** [reply]
Hi @taylor13,

I can confirm that the xml files in the "current" directories are indeed the controlled vocabularies. The region names and area types are structured very similarly to the xml for the standard name table, except they have no units tag. The html versions of the table are rendered from these xml files for ease of browsing.

**Comment by taylor13** [comment]
I'm curious why "xml" is perferred to "json" as a format.  I've always found the json files easier for humans to read and also trivial to ingest using python.  I suppose there is some historical reason.

**Reply by japamment** [reply]
The xml files are the result of a gradual evolution of the vocabularies, in terms of both content and formatting. JSON-LD versions are available from the NERC Vocabulary Server (NVS) maintained by @gwemon and team.

**A brief history of the xml files**
For a long time (2002 - 2013) we had only html versions of the area types and standardized regions. They were simple lists of terms and had no accompanying descriptions. In fact, we still maintain the [old html file](https://cfconventions.org/Data/cf-standard-names/docs/standardized-region-names.html) for the regions as it has some additional information in the header. As we added more area types (many requested for phases of CMIP), the need to add descriptions became more important as a way to explain and disambiguate the terms. About 10 years ago I created xml files using a schema that was based on the standard name table schema in Appendix B of the CF conventions. The main consideration was to create a means to store the terms and descriptions together. We don't have many descriptions for the regions because generally they aren't needed, but there are one or two to disambiguate similar terms. Using a schema similar to the standard names was a pragmatic decision because it was simple to then adapt the processing scripts which render the vocabularies as html for browsing. It's also worth mentioning that, prior to the creation of the xml files, the region list and area types grew organically without any formal versioning. Since 2013, the xml files have been maintain with numbered versions and date stamps analogous to the process for standard names. We don't currently have any aliases of area types or regions, but if they were ever needed we could handle them (and serve them) just as we do for standard names.

**Availability of JSON files**
Since 2006 the standard names have been published in NVS as collection [P07](https://vocab.nerc.ac.uk/collection/P07/current/) and in 2013 the area types and standardized regions were added as collections [P29](https://vocab.nerc.ac.uk/collection/P29/current/) and [P30](https://vocab.nerc.ac.uk/collection/P30/current/) respectively. In all three cases, there are links at the top right of the page to view the collection in RDF/XML, Turtle and JSON-LD formats.

In the [detailed discussion](https://github.com/orgs/cf-convention/discussions/366) of how we can make greater use of NVS for the benefit of the CF community, I see you have suggested a JSON file containing only the CF defined properties for standard names (name, canonical units, aliases). I've no doubt we could construct this, however I think it's best if we continue to talk about that in [#366](https://github.com/orgs/cf-convention/discussions/366). For me, it's a question of whether we should be maintaining even more formats of files in the CF GitHub or whether your requirements can be met via NVS, perhaps by having a CF profile that would provide access to a subset of the information.
